pkg/jobs/metricspoller/job_statistics.go

```diff
@@ -100,21 +100,27 @@ type schedulePTSStat struct {
 func manageProtectedTimestamps(ctx context.Context, execCtx sql.JobExecContext) error {
 	var ptsStats map[jobspb.Type]*ptsStat
 	var schedulePtsStats map[string]*schedulePTSStat
-
+	var ptsState ptpb.State
 	execCfg := execCtx.ExecCfg()
+
 	if err := execCfg.InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
 		ptsStats = make(map[jobspb.Type]*ptsStat)
 		schedulePtsStats = make(map[string]*schedulePTSStat)
+		var err error
+		ptsState, err = execCfg.ProtectedTimestampProvider.WithTxn(txn).GetState(ctx)
+		return err
+	}); err != nil {
+		return err
+	}
 
-		ptsState, err := execCfg.ProtectedTimestampProvider.WithTxn(txn).GetState(ctx)
+	// Iterate over each job with a fresh transaction, to prevent scanning the
+	// whole jobs table with a single txn.
+	for _, rec := range ptsState.Records {
+		id, err := jobsprotectedts.DecodeID(rec.Meta)
 		if err != nil {
 			return err
 		}
-		for _, rec := range ptsState.Records {
-			id, err := jobsprotectedts.DecodeID(rec.Meta)
-			if err != nil {
-				return err
-			}
+		if err := execCfg.InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
 			switch rec.MetaType {
 			case jobsprotectedts.GetMetaType(jobsprotectedts.Jobs):
```

Comment:
I wonder if inside this transaction we should call `execCfg.ProtectedTimestampProvider.WithTxn(txn).GetRecord` to ensure that the PTS is still too old in the same transaction where we decide to cancel the job. 

Although, I'm willing to forgo this if you feel strongly since we know that there was some point in time at which this job should have been canceled and we only give up a very narrow case in which the job just happened to catch up.

Reply:
yeah, I'm fine rechecking this.